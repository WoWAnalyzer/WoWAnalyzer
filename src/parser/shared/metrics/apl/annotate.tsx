import { SpellLink } from 'interface';

import type { Tense, CheckResult, InternalRule, Violation } from './index';
import { addInefficientCastReason } from 'parser/core/EventMetaLib';
import { Fragment } from 'react';

export function ConditionDescription({
  tense,
  rule,
  prefix,
}: {
  rule: InternalRule;
  prefix?: string;
  tense?: Tense;
}) {
  if (!('condition' in rule)) {
    return null;
  }

  const desc = rule.condition?.describe(tense);
  const customPrefix = rule.condition?.prefix ?? prefix ?? 'because';

  if (!desc || desc === '') {
    return null;
  }

  return (
    <>
      {' '}
      {customPrefix}
      {customPrefix ? ' ' : ''}
      {desc}
    </>
  );
}

function InefficientCastAnnotation({ violation }: { violation: Violation }) {
  return (
    <>
      {violation.expectedCast.map((spell, index) => (
        <Fragment key={spell.id}>
          {index > 0 ? ' and ' : ''}
          <SpellLink spell={spell.id} />
        </Fragment>
      ))}{' '}
      {violation.expectedCast.length > 1 ? 'were' : 'was'} available and higher priority
      <ConditionDescription rule={violation.rule} />.
    </>
  );
}

/**
 * Annotate a list of APL violations on the timeline.
 *
 * This mutates the input and will overwrite other annotations, but is
 * idempotent. There is no return value.
 */
export default function annotateTimeline(violations: CheckResult['violations']) {
  for (const violation of violations) {
    addInefficientCastReason(
      violation.actualCast,
      <InefficientCastAnnotation violation={violation} />,
    );
  }
}
