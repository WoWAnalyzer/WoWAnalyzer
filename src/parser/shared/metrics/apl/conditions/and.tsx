import type { Condition } from '../index';
import { containsOptionalCondition } from './optionalRule';

type ConditionMap = { [k: string]: Condition<any> };

export default function and(...conditions: Array<Condition<any>>): Condition<any> {
  const key = `and-${conditions.map((cnd) => cnd.key).join('-')}`;
  if (!import.meta.env.PROD && conditions.some(containsOptionalCondition)) {
    console.warn(
      `APL rule ${key} contains optional rules. Nesting optionalRule inside of and can produce confusing behavior and is discouraged.`,
      conditions,
    );
  }
  const cndMap: ConditionMap = conditions.reduce((map: ConditionMap, cnd) => {
    map[cnd.key] = cnd;
    return map;
  }, {});

  const lookahead = conditions.reduce(
    (val: number | undefined, cnd) => (!val || (cnd.lookahead || 0) > val ? cnd.lookahead : val),
    undefined,
  );

  return {
    key,
    init: (info) => Object.fromEntries(conditions.map(({ key, init }) => [key, init(info)])),
    update: (state, event) =>
      Object.keys(state).reduce((nextState: ConditionMap, key) => {
        nextState[key] = cndMap[key]!.update(state[key], event);
        return nextState;
      }, {}),
    lookahead,
    validate: (state, event, spell, lookahead) =>
      Object.keys(state).every((key) => cndMap[key]!.validate(state[key], event, spell, lookahead)),
    describe: (tense) =>
      conditions
        .map((cnd) => cnd.describe(tense))
        .reduce((cur, next) =>
          cur ? (
            <>
              {cur} and {next}
            </>
          ) : (
            next
          ),
        ),
  };
}
