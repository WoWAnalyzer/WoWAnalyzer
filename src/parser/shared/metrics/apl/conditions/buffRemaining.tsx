import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import {
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  EventType,
  RefreshBuffEvent,
  RemoveBuffEvent,
} from 'parser/core/Events';
import { Condition, tenseAlt } from '../index';
import { buffDuration, DurationData, formatTimestampRange, PandemicData, Range } from './util';

type State = DurationData & {
  previousEvent: ApplyBuffEvent | RefreshBuffEvent | ApplyBuffStackEvent | RemoveBuffEvent;
};

/**
 * Assert that a buff has remaining time within `range`.
 */
export default function buffRemaining(
  spell: Spell,
  pandemic: PandemicData,
  range: Range,
  /**
   * If true, then if the buff is missing it will be considered valid if only atMost is specified
   */
  missingCanBeValid: boolean = false,
): Condition<State | undefined> {
  return {
    key: `buffRemaining-${spell.id}`,
    init: () => undefined,
    update: (state, event) => {
      if (event.type === EventType.ApplyBuff && event.ability.guid === spell.id) {
        return {
          referenceTime: event.timestamp,
          previousEvent: event,
          timeRemaining: pandemic.duration,
        };
      } else if (
        (event.type === EventType.RefreshBuff || event.type === EventType.ApplyBuffStack) &&
        event.ability.guid === spell.id
      ) {
        // *sometimes* the game logs both a refresh and an applybuffstack at the same timestamp.
        // this bit handles not extending the buff twice when that occurs.
        if (
          state &&
          state.previousEvent.type !== event.type &&
          state.previousEvent.timestamp === event.timestamp
        ) {
          return state;
        }
        return {
          referenceTime: event.timestamp,
          previousEvent: event,
          timeRemaining: buffDuration(event.timestamp, state, pandemic),
        };
      } else if (event.type === EventType.RemoveBuff && event.ability.guid === spell.id) {
        return undefined;
      } else {
        return state;
      }
    },
    validate: (state, event) => {
      if (state === undefined) {
        //Buff is not applied so buff has no remaining time so only atMost can be true
        return Boolean(missingCanBeValid && range.atMost && !range.atLeast);
      }
      const timeRemaining = buffDuration(event.timestamp, state, pandemic);
      return (
        timeRemaining >= (range.atLeast || 0) && (!range.atMost || timeRemaining <= range.atMost)
      );
    },
    describe: (tense) => (
      <>
        <SpellLink spell={spell.id} /> {tenseAlt(tense, 'has', 'had')} {formatTimestampRange(range)}{' '}
        seconds remaining
      </>
    ),
  };
}
