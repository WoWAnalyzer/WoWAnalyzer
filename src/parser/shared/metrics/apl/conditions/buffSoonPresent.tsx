import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import { EventType } from 'parser/core/Events';
import { Condition, tenseAlt } from '../index';
import { formatTimestampRange, Range } from './util';
import { TrackedBuffEvent } from 'parser/core/Entity';

type BuffInformation = {
  buffs: TrackedBuffEvent[];
  nextBuff: number | undefined;
  currentIndex: number;
};

type SourceOptions = 'player' | 'external' | 'any';

/**
 * Check if a buff will be present soon after the present time.
 *
 * This should ONLY be used for buffs that the player has complete control over, such as buffs tied to cooldown use.
 * DO NOT use this to "look ahead" for random procs to implement impossibly optimal APLs.
 */
export function buffSoonPresent(
  spell: Spell,
  range: Range,
  /** Set to configure the source of the buff */
  source: SourceOptions = 'player',
): Condition<BuffInformation> {
  return {
    key: `buffSoonPresent-${spell.id}-${range.atLeast}-${range.atMost}-${source}`,
    init: ({ combatant }) => {
      const buffs = combatant?.buffs
        ? combatant.buffs.filter(
            (buff) =>
              buff.ability.guid === spell.id && isValidSource(source, combatant.id, buff.sourceID),
          )
        : [];

      return {
        buffs,
        nextBuff: buffs[0]?.timestamp,
        currentIndex: 0,
      };
    },
    update: (state, event) => {
      if (event.type !== EventType.ApplyBuff || event.ability.guid !== spell.id) {
        return state;
      }

      state.currentIndex += 1;
      state.nextBuff = state.buffs[state.currentIndex]?.timestamp;

      return state;
    },
    validate: (state, event) => {
      /** No more buffs or no buffs found.
       * We will assume no more buffs will come so only atLeast can be true */
      if (!state.nextBuff) {
        return Boolean(!range.atMost && range.atLeast);
      }

      const timeTillNextBuff = state.nextBuff - event.timestamp;
      return (
        timeTillNextBuff >= (range.atLeast || 0) &&
        (!range.atMost || timeTillNextBuff <= range.atMost)
      );
    },
    describe: (tense) => (
      <>
        there {tenseAlt(tense, 'is', 'was')} {formatTimestampRange(range)} seconds remaining before{' '}
        <SpellLink spell={spell.id} /> {tenseAlt(tense, 'is', 'was')} present
      </>
    ),
  };
}

function isValidSource(source: SourceOptions, playerId: number, buffSourceId?: number): boolean {
  switch (source) {
    case 'any':
      return true;
    case 'player':
      return buffSourceId === playerId;
    case 'external':
      return buffSourceId !== playerId;
    default:
      return false;
  }
}
