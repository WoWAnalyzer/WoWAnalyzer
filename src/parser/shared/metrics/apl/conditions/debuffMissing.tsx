import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import { EventType, GetRelatedEvents, HasTarget } from 'parser/core/Events';
import { encodeEventTargetString, encodeTargetString } from 'parser/shared/modules/Enemies';

import { AplTriggerEvent, Condition, tenseAlt } from '../index';
import { validateBuffMissing } from './buffPresent';
import { buffDuration, DurationData, PandemicData } from './util';

export type TargetOptions = {
  /**
   * Link relation (see `EventLinkNormalizer`) to use for determining
   * targets on untargeted debuff-applying spells.
   *
   * When present, overrides any targets the event has.
   */
  targetLinkRelation?: string;
};

export function getTargets(event: AplTriggerEvent, targetLink?: string): string[] {
  if (targetLink) {
    return GetRelatedEvents(event, targetLink)
      .map(encodeEventTargetString)
      .filter((key): key is string => key !== null);
  } else if (HasTarget(event)) {
    return [encodeEventTargetString(event)!];
  } else {
    return [];
  }
}

/**
   The rule applies when the debuff `spell` is missing. The `optPandemic`
   parameter gives the ability to allow early refreshes to prevent a debuff
   dropping, but this will not *require* early refreshes.

   fallback defines the output when there is no target/the debuff hasn't been seen yet
 **/
export function debuffMissing(
  spell: Spell,
  optPandemic?: PandemicData,
  targetOptions?: TargetOptions,
  fallback: boolean = false,
): Condition<{ [key: string]: DurationData }> {
  return {
    key: `debuffMissing-${spell.id}`,
    init: () => ({}),
    update: (state, event) => {
      switch (event.type) {
        case EventType.RefreshDebuff:
        case EventType.ApplyDebuff:
          if (event.ability.guid === spell.id) {
            const encodedTargetString = encodeTargetString(event.targetID, event.targetInstance);
            state[encodedTargetString] = {
              referenceTime: event.timestamp,
              timeRemaining: optPandemic
                ? buffDuration(event.timestamp, state[encodedTargetString], optPandemic)
                : undefined,
            };

            return state;
          }
          break;
        case EventType.RemoveDebuff:
          if (event.ability.guid === spell.id) {
            const encodedTargetString = encodeTargetString(event.targetID, event.targetInstance);
            delete state[encodedTargetString];
            return state;
          }
          break;
      }

      return state;
    },
    validate: (state, event, ruleSpell) => {
      const targets = getTargets(event, targetOptions?.targetLinkRelation);
      if (targets.length === 0) {
        //No target so we can't check for a debuff
        return fallback;
      }

      return targets.some((encodedTargetString) =>
        validateBuffMissing(event, ruleSpell, state[encodedTargetString], optPandemic),
      );
    },
    describe: (tense) => (
      <>
        <SpellLink spell={spell.id} /> {tenseAlt(tense, 'is', 'was')} missing{' '}
        {optPandemic && <>or about to expire</>} from the target
      </>
    ),
  };
}
