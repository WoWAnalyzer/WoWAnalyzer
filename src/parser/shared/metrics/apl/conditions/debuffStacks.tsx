import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';
import { EventType } from 'parser/core/Events';

import { Condition, tenseAlt } from '../index';
import { Range, formatRange } from './index';

export default function debuffStacks(spell: Spell, range: Range): Condition<number> {
  return {
    key: `debuffStacks-${spell.id}`,
    init: () => 0,
    update: (state, event) => {
      switch (event.type) {
        case EventType.ApplyDebuffStack:
          if (event.ability.guid === spell.id) {
            return event.stack;
          }
          break;
        case EventType.ApplyDebuff:
          if (event.ability.guid === spell.id) {
            return 1;
          }
          break;
        case EventType.RemoveDebuffStack:
          if (event.ability.guid === spell.id) {
            return event.stack;
          }
          break;
        case EventType.RemoveDebuff:
          if (event.ability.guid === spell.id) {
            return 0;
          }
          break;
      }
      return state;
    },
    validate: (state, _event) =>
      state >= (range.atLeast || 0) && (range.atMost === undefined || state <= range.atMost),
    describe: (tense) => (
      <>
        enemie {tenseAlt(tense, 'have', 'had')} {formatRange(range)}{' '}
        <SpellLink spell={spell.id} icon /> stacks
      </>
    ),
  };
}
