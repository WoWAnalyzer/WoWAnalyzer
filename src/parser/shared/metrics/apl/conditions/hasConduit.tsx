import type Spell from 'common/SPELLS/Spell';
import { SpellLink } from 'interface';

import { Condition, tenseAlt } from '../index';

/**
 * @deprecated
 */
export default function hasConduit(conduit: Spell): Condition<boolean> {
  return {
    key: `hasConduit-${conduit.id}`,
    init: () => false,
    update: (state) => state,
    validate: (state) => state,
    describe: (tense) => (
      <>
        you {tenseAlt(tense, 'have', 'had')} <SpellLink id={conduit.id} /> equipped
      </>
    ),
  };
}
