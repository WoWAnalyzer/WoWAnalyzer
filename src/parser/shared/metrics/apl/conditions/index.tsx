export { default as hasConduit } from './hasConduit';
export { default as hasLegendary } from './hasLegendary';
export { default as optionalRule } from './optionalRule';
export { buffPresent, buffMissing } from './buffPresent';
export { default as hasTalent } from './hasTalent';
export { default as spellAvailable } from './spellAvailable';
export { default as spellCooldownRemaining } from './spellCooldownRemaining';
export { default as buffRemaining } from './buffRemaining';
export { default as hasResource } from './hasResource';
export { default as inExecute } from './inExecute';
export { default as or } from './or';
export { default as not } from './not';
export { default as always } from './always';
export { default as and } from './and';
export { default as spellCharges } from './spellCharges';
export { spellFractionalCharges } from './spellFractionalCharges';
export { default as buffStacks } from './buffStacks';
export { default as targetsHit } from './targetsHit';
export { debuffPresent } from './debuffPresent';
export { debuffMissing } from './debuffMissing';
export { hasCovenant } from './hasCovenant';
export { lastSpellCast } from './lastSpellCast';
export { default as describe } from './describe';
export * from './util';
export { has2PieceByTier, has4PieceByTier } from './hasTier';
export { buffSoonPresent } from './buffSoonPresent';
