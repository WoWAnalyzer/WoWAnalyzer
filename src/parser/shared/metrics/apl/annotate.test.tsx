import renderer from 'react-test-renderer';
import { Provider as ReduxProvider } from 'react-redux';
import { store } from 'store';

import { ConditionDescription } from './annotate';
import { buffPresent } from './conditions';

describe('ConditionDescription', () => {
  it('should return no description for unconditional rules', () => {
    const content = renderer
      .create(
        <ReduxProvider store={store}>
          <ConditionDescription rule={{ id: 1, name: 'Test', icon: '' }} />
        </ReduxProvider>,
      )
      .toJSON();
    expect(content).toBeNull();
  });

  it('should return a description for a rule with a condition', () => {
    const rule = {
      spell: { id: 1, name: 'Test', icon: '' },
      condition: buffPresent({ id: 2, name: 'Buff', icon: '' }),
    };

    const content = renderer
      .create(
        <ReduxProvider store={store}>
          <ConditionDescription rule={rule} />
        </ReduxProvider>,
      )
      .toJSON();

    expect(content).toMatchSnapshot();
  });
});
