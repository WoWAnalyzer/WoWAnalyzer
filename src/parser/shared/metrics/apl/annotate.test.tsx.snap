// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConditionDescription > should return a description for a rule with a condition 1`] = `
[
  " ",
  "because",
  " ",
  <a
    className="spell-link-text"
    href="https://wowhead.com/spell=2"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt=""
      className="icon game "
      src="https://assets.rpglogs.com/img/warcraft/abilities/inv_misc_questionmark.jpg"
    />
     
    Unknown spell: 2
  </a>,
  " ",
  "was",
  " present",
]
`;
